### PR TITLE
Grazing

### PR DESCRIPTION
**Pull request by ckoven**
An initial implementation to allow herbivores to graze and browse plants, with different grazing rates on different land use classes.

### Description:
This PR adds the basic capability to have grazers and browsers eat plants. The basic idea is to specify this as a rate of consumption in units of 1/time. I.e. grazers and browsers eat X% of all accessible leaves of edible plants per day on columns of land-use type Y, as controlled by the parameter `fates_landuse_grazing_rate(fates_landuseclass)`. The specification of what is considered edible plants is parametric and controlled by a PFT-level trait `fates_landuse_grazing_palatability(fates_pft)`, so e.g. this can be specified to include grazers (who eat herbaceous plants only) and/or browsers (who eat the leaves of woody plants). And then all leaves of that PFT on columns of that land use type that are below a height of parameter `fates_landuse_grazing_maxheight` can be subject to herbivory.

The fate of the eaten material is controlled by three "use-efficiency" parameters: `fates_landuse_grazing_carbon_use_eff`', `fates_landuse_grazing_nitrogen_use_eff`, 'fates_landuse_grazing_phosphorus_use_eff'. So some mass can be lost from the ecosystem and some can be added to the litter pools via manure (and, I suppose, mortality of grazers).

Default parameter values that are there now (all of which extremely subject to change) are to make this only apply on pasture and rangeland, with a rate of 4% per day based on https://doi.org/10.5194/gmd-11-815-2018 by @samsrabin et al., and currently only grazers (i.e. only applies to nonwoody PFTs). I couldn't find a whole lot of data on the use-efficiency parameters, but those shouldn't be too important until we are running nutrient cycles at large scale under transient change.

fixes https://github.com/NGEET/fates/discussions/936

### Expectation of Answer Changes:
This will be answer changing if the parameters are set to allow grazing.

### Checklist

*If this is your first time contributing, please read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/main/CONTRIBUTING.md) document.*

All checklist items must be checked to enable merging this pull request:

*Contributor*
- [x] The in-code documentation has been updated with descriptive comments
- [x] The documentation has been assessed to determine if updates are necessary

*Integrator*
- [ ] FATES PASS/FAIL regression tests were run
- [ ] Evaluation of test results for answer changes was performed and results provided

### Documentation

- [Technical Note](https://github.com/NGEET/fates-docs) update:
- [User's Guide](https://github.com/NGEET/fates-users-guide) update: 

### Test Results:

*CTSM (or) E3SM (specify which) test hash-tag:*

*CTSM (or) E3SM (specify which) baseline hash-tag:*

*FATES baseline hash-tag:*

*Test Output:*

### CODE REVIEW

File: parteh/PRTGenericMod.F90 [hunk 1]
```diff
@@ -1124,6 +1133,39 @@ function GetBurned(this, organ_id, element_id, position_id) result(burned_val)
       return
    end function GetBurned
 
+    ! =========================================================================
+    
+    function GetHerbivory(this, organ_id, element_id, position_id) result(herbivory_val)
+
+      ! This function is very very similar to GetBurned
```
Comment (samsrabin) [inline parteh/PRTGenericMod.F90:1140]: In fact it looks identical, aside from `herbivory` replacing `burned`. Could these be replaced with a single function? Named, e.g., `GetDisturbed`.

File: parteh/PRTLossFluxesMod.F90 [hunk 2]
```diff
@@ -877,7 +878,62 @@ subroutine PRTDamageRecoveryFluxes(prt, organ_id, mass_0, mass, cc_mass)
...
+          
+          i_var = organ_map(organ_id)%var_id(i_var_of_organ)
+          
+          element_id = prt_global%state_descriptor(i_var)%element_id
+          
+          ! Loop over all of the coordinate ids
+          do i_pos = 1,prt_global%state_descriptor(i_var)%num_pos
+             
+             ! The mass that is leaving the plant
+             consumed_mass = mass_fraction * prt%variables(i_var)%val(i_pos)
```
Comment (samsrabin) [inline parteh/PRTLossFluxesMod.F90:921]: `consumed_mass` should presumably be `herbivore_consumed_mass`?

**Review by samsrabin — COMMENTED** [review]
Not sure if this was actually ready for review, as it has the "draft" tag but isn't a Draft in the Github sense. Anyway, I've made some (initial?) comments. Also: Where are the new parameters actually read?

**Comment by ckoven** [comment]
> Not sure if this was actually ready for review, as it has the "draft" tag but isn't a Draft in the Github sense. Anyway, I've made some (initial?) comments. Also: Where are the new parameters actually read?

Thanks @samsrabin! Yes, this is draft in the sense that it wasn't done or tested. And yes, one of the things that isn't done yet is the code to actually read the new parameters. But better to get eyes on it sooner rather than later.

File: main/EDMainMod.F90 [hunk 3]
```diff
@@ -524,6 +525,9 @@ subroutine ed_integrate_state_variables(currentSite, bc_in, bc_out )
              bc_out%ar_site = bc_out%ar_site + currentCohort%resp_acc_hold * & 
                   AREA_INV * currentCohort%n / hlm_days_per_year / sec_per_day
 
+             ! allow herbivores to graze
+             call fates_grazing(currentCohort%prt, ft, currentPatch%land_use_label, currentCohort%height)
```
Comment (glemieux) [inline main/EDMainMod.F90:529]:
```suggestion
             call FatesGrazing(currentCohort%prt, ft, currentPatch%land_use_label, currentCohort%height)
```

File: main/FatesHistoryInterfaceMod.F90 [hunk 4]
```diff
@@ -2891,6 +2894,9 @@ subroutine update_history_dyn1(this,nc,nsites,sites,bc_in)
                   hio_npp_stor_si(io_si) = hio_npp_stor_si(io_si) +               &
                        store_m_net_alloc * n_perm2 / days_per_year / sec_per_day
 
+                  leaf_herbivory   = ccohort%prt%GetHerbivory(leaf_organ, carbon12_element) * days_per_year  !cdkcdk
+                  hio_grazing_si(io_si) = hio_grazing_si(io_si) + leaf_herbivory * n_perm2 / days_per_year / sec_per_day
```
Comment (glemieux) [inline main/FatesHistoryInterfaceMod.F90:2898]: I've never really thought about this: Why do we multiply by `days_per_year` and then divide it out again?

**Review by glemieux — CHANGES_REQUESTED** [review]
I had some minor suggested name changes and one question.  I'll push an update to add the xml patch file for the cdl update.

**Comment by glemieux** [comment]
Preliminary regression testing against `sci.1.80.6_api.37.0.0` baseline shows expected FIELDLIST differences due to the new `FATES_GRAZING` history output not being available in the baselines.  Almost all other tests are B4B, with the exception of the `FatesColdPRT2` test:

```
  525 dec1212.hsn.de.hpc.ucar.edu 622: forrtl: severe (408): fort: (2): Subscript #1 of the array MASS_BALANCE has value 4 which is greater than the upper bound of 3
  526 dec1212.hsn.de.hpc.ucar.edu 622:
  527 dec1212.hsn.de.hpc.ucar.edu 622: Image              PC                Routine            Line        Source
  528 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001D15A0B  edphysiologymod_m        2964  EDPhysiologyMod.F90
  529 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001CD062B  edphysiologymod_m         485  EDPhysiologyMod.F90
  530 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001BFE5BB  edmainmod_mp_ed_i         785  EDMainMod.F90
  531 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001BF7DB2  edmainmod_mp_ed_e         223  EDMainMod.F90
  532 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000000B13944  clmfatesinterface        1250  clmfates_interfaceMod.F90
  533 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000000A89E08  clm_driver_mp_clm        1143  clm_driver.F90
```

I haven't dug into this yet.  Any initial thoughts @rgknox?

**Comment by ckoven** [comment]
> Preliminary regression testing against `sci.1.80.6_api.37.0.0-ctsm5.3.018` baseline shows expected FIELDLIST differences due to the new `FATES_GRAZING` history output not being available in the baselines. Almost all other tests are B4B, with the exception of the `FatesColdPRT2` test:
> 
> ```
>   525 dec1212.hsn.de.hpc.ucar.edu 622: forrtl: severe (408): fort: (2): Subscript #1 of the array MASS_BALANCE has value 4 which is greater than the upper bound of 3
>   526 dec1212.hsn.de.hpc.ucar.edu 622:
>   527 dec1212.hsn.de.hpc.ucar.edu 622: Image              PC                Routine            Line        Source
>   528 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001D15A0B  edphysiologymod_m        2964  EDPhysiologyMod.F90
>   529 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001CD062B  edphysiologymod_m         485  EDPhysiologyMod.F90
>   530 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001BFE5BB  edmainmod_mp_ed_i         785  EDMainMod.F90
>   531 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000001BF7DB2  edmainmod_mp_ed_e         223  EDMainMod.F90
>   532 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000000B13944  clmfatesinterface        1250  clmfates_interfaceMod.F90
>   533 dec1212.hsn.de.hpc.ucar.edu 622: cesm.exe           0000000000A89E08  clm_driver_mp_clm        1143  clm_driver.F90
> ```
> 
> I haven't dug into this yet. Any initial thoughts @rgknox?
> 
> Results: `/glade/u/home/glemieux/scratch/ctsm-tests/tests_pr1140-fates-woodyfix`

I think I see the error that caused this, will fix.

File: biogeochem/FatesLandUseChangeMod.F90 [hunk 5]
```diff
@@ -424,4 +425,44 @@ subroutine GetInitLanduseTransitionRates(bc_in, min_allowed_landuse_fraction, &
     
   end subroutine GetInitLanduseTransitionRates
     
+  !----------------------------------------------------------------------------------------------------
+
+  subroutine fates_grazing(prt, ft, land_use_label, height)
```
Comment (ckoven) [inline biogeochem/FatesLandUseChangeMod.F90:430]:
```suggestion
  subroutine FatesGrazing(prt, ft, land_use_label, height)
```

File: biogeochem/FatesLandUseChangeMod.F90 [hunk 6]
```diff
@@ -424,4 +425,44 @@ subroutine GetInitLanduseTransitionRates(bc_in, min_allowed_landuse_fraction, &
...
+          
+          grazing_rate = grazing_rate * &
+               max(0._r8, min(1._r8, &
+               (landuse_grazing_maxheight - (height - crown_depth )) / crown_depth))
+       endif
+
+       call PRTHerbivoryLosses(prt, leaf_organ, grazing_rate)
+    end if
+
+  end subroutine fates_grazing
```
Comment (ckoven) [inline biogeochem/FatesLandUseChangeMod.F90:466]:
```suggestion
  end subroutine FatesGrazing
```

File: main/EDMainMod.F90 [hunk 7]
```diff
@@ -105,6 +105,7 @@ module EDMainMod
   use PRTParametersMod      , only : prt_params
   use EDPftvarcon,            only : EDPftvarcon_inst
   use FatesHistoryInterfaceMod, only : fates_hist
+  use FatesLandUseChangeMod,  only: fates_grazing
```
Comment (ckoven) [inline main/EDMainMod.F90:108]:
```suggestion
  use FatesLandUseChangeMod,  only: FatesGrazing
```

File: biogeochem/EDPhysiologyMod.F90 [hunk 8]
```diff
@@ -2849,6 +2851,8 @@ subroutine CWDInput( currentSite, currentPatch, litt, bc_in)
     integer  :: numlevsoil        ! Actual number of soil layers
 
     real(r8) :: SF_val_CWD_frac_adj(4) !SF_val_CWD_frac adjusted based on cohort dbh
+    real(r8) :: leaf_herbivory
+    real(r8) :: herbivory_element_use_efficiency
     !----------------------------------------------------------------------
```
Comment (jenniferholm) [inline biogeochem/EDPhysiologyMod.F90:2856]: Maybe give a short definition of these two? leaf_herbivory and herbivory_element_use_efficiency

File: biogeochem/EDPhysiologyMod.F90 [hunk 9]
```diff
@@ -2928,7 +2943,9 @@ subroutine CWDInput( currentSite, currentPatch, litt, bc_in)
        do dcmpy=1,ndcmpy
           dcmpy_frac = GetDecompyFrac(pft,leaf_organ,dcmpy)
           litt%leaf_fines_in(dcmpy) = litt%leaf_fines_in(dcmpy) + &
-               (leaf_m_turnover+repro_m_turnover) * plant_dens * dcmpy_frac
+               (leaf_m_turnover+repro_m_turnover + &
+               leaf_herbivory * herbivory_element_use_efficiency) * &
```
Comment (jenniferholm) [inline biogeochem/EDPhysiologyMod.F90:2947]: Just to help me understand the steps, will the leaf litter here already take into account the leaf mass that is consumed and "stays" in the animal as food, and only a portion returning to the litter as "manure"?  Is that the herbivory_element_use_efficiency parameter, or something else?  Because I'm also confused about what portion of the litter is food, is manure, and/or to the atmosphere, and if all 3 paths are tracked?

File: biogeochem/EDPhysiologyMod.F90 [hunk 10]
```diff
@@ -2941,6 +2958,11 @@ subroutine CWDInput( currentSite, currentPatch, litt, bc_in)
             elflux_diags%root_litter_input(pft) +  &
             (fnrt_m_turnover + store_m_turnover ) * currentCohort%n
 
+       ! send the part of the herbivory flux that doesn't go to litter to the atmosphere
+
+       site_mass%herbivory_flux_out = &
+            site_mass%herbivory_flux_out + &
+            leaf_herbivory * (1._r8 - herbivory_element_use_efficiency) * currentCohort%n
```
Comment (jenniferholm) [inline biogeochem/EDPhysiologyMod.F90:2965]: Since it looks like 1._r8 - herbivory_element_use_efficiency is the portion that goes to the atmosphere, is the other portion only going to manure, food diet consumption, or both?  My bad if I'm not following

File: main/FatesHistoryInterfaceMod.F90 [hunk 11]
```diff
@@ -6764,6 +6770,12 @@ subroutine define_history_vars(this, initialize_variables)
             upfreq=group_dyna_simple, ivar=ivar, initialize=initialize_variables,                &
             index = ih_npp_stor_si)
 
+       call this%set_history_var(vname='FATES_GRAZING', units='kg m-2 s-1',    &
+            long='grazing by herbivores of leaves in kg carbon per m2 per second',          &
```
Comment (jenniferholm) [inline main/FatesHistoryInterfaceMod.F90:6774]: This is great to have this FATES_GRAZING output and know the mass of leaves eaten.  In the future could we also have a FATES_GRAZING_LITTER_OUT type output, that gives us an idea of the amount of manure output. Or helps to differentiate how much litter is natural from turnover vs. litter from herbivores.  And maybe useful later for CH4 tracking?

File: parameter_files/fates_params_default.cdl [hunk 12]
```diff
@@ -1769,6 +1791,14 @@ data:
 
  fates_hydro_solver = 1 ;
 
+ fates_landuse_grazing_carbon_use_eff = 0. ;
+
+ fates_landuse_grazing_maxheight =  1 ;
+
+ fates_landuse_grazing_nitrogen_use_eff = 0.25 ;
+
+ fates_landuse_grazing_phosphorus_use_eff = 0.5 ;
```
Comment (jenniferholm) [inline parameter_files/fates_params_default.cdl:1800]: Are these _use_eff values just placeholders, or from literature (Sam's paper)?  Why again is the carbon_use_eff = 0 ?

File: parteh/PRTLossFluxesMod.F90 [hunk 13]
```diff
@@ -885,7 +886,62 @@ subroutine PRTDamageRecoveryFluxes(prt, organ_id, mass_0, mass, cc_mass)
...
+     ! =====================================================================================
+
+  subroutine PRTHerbivoryLosses(prt, organ_id, mass_fraction)
+
+    ! ----------------------------------------------------------------------------------
+    ! This subroutine assumes that there is no re-translocation associated
+    ! with herbivory. There is only one destiny for eaten mass within
+    ! the organ, and that is outside the plant.  
+    ! It is also assumed that non PARTEH parts of the code (ie the grazing-model)
+    ! will decide what to do with the consumed mass (i.e. sent it to the litter
```
Comment (jenniferholm) [inline parteh/PRTLossFluxesMod.F90:899]: Just for clarity and ease of understanding, maybe explicitly say where the “grazing-model” is for tracking and clarity. Is this “subroutine FatesGrazing” you mean or something else?

**Review by jenniferholm — COMMENTED** [review]
Thanks Charlie this is really cool work.  My biggest question was just some confusion around if mass can be loss from the ecosystem via both food/consumption pathway and to the litter flux to the atmosphere, and added to the litter via manure. Are all three pathways occurring, and is the order correct?  Plus does litter from natural litter fluxes plus manure equal the total litter in the history outputs?  I'm not sure that made sense, but wondering if we could separate those out?  Thanks!

File: parteh/PRTLossFluxesMod.F90 [hunk 14]
```diff
@@ -885,7 +886,62 @@ subroutine PRTDamageRecoveryFluxes(prt, organ_id, mass_0, mass, cc_mass)
...
+    ! pool or send to atmosphere)
```
Comment (ckoven) [inline parteh/PRTLossFluxesMod.F90:900]:
```suggestion
    ! This subroutine assumes that there is no re-translocation associated
    ! with herbivory. There are two pathways for grazed matter. Some fraction 
    ! is lost from the land system (i.e. to the atmosphere via respiration or gas-phase 
    ! nutrient loss pathways), and some fraction is passed to litter decomposition 
    ! (i.e. via manure or grazer mortality). These pathways of consumed biomass 
    ! are handled outside PARTEH (e.g., the litter flux is handled in EDPhysiologyMod, 
    ! and controlled by element-specific "use efficiency" parameters).
```

**Review by ckoven — COMMENTED** [review]
Thanks @jenniferholm for these questions. Tried to update the documentation to capture these points.

**Comment by glemieux** [comment]
> Preliminary regression testing against `sci.1.80.6_api.37.0.0-ctsm5.3.018` baseline shows expected FIELDLIST differences due to the new `FATES_GRAZING` history output not being available in the baselines. Almost all other tests are B4B, with the exception of the `FatesColdPRT2` test:

Rerunning the full `fates` regression tests results in all passing now.  Note that these were not tested against a baseline as these are intended as preliminary.

File: parameter_files/archive/api37.0.0_011025_patch_params.xml [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+      <ln> Relative intensity of leaf grazing/browsing per PFT </ln>
+      <val> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1 </val>
+    </mod>
+    <mod type="variable_add">
+      <na> fates_landuse_grazing_rate </na>
+      <di> fates_landuseclass </di>
+      <un> 1/day </un>
+      <ln> fraction of leaf biomass consumed by grazers per day </ln>
+      <val> 0.0, 0.0, 0.0, 0.0, 0.0 </val>
+    </mod>
```
Comment (glemieux) [inline parameter_files/archive/api37.0.0_011025_patch_params.xml:38]: @ckoven we should add a fates grazing test that checks non-zero grazing rates at some point in the near future.  Can you recommend some  parameter values for the testing?

**Review by glemieux — APPROVED** [review]

**Comment by glemieux** [comment]
Final regression testing last week is returning B4B results as expected against `fates-sci.1.80.14_api.37.0.0-ctsm5.3.024`.  Only differences showing are due to the default parameter update and `FIELDLIST` differences for the new grazing output added.
